### PR TITLE
Modifying Dockerfile.amd64  to use debian instead of alpine as base

### PR DESCRIPTION
**Pull request by stevegaossou**
## Description
Modifying Dockerfiles to use debian instead of alpine as base for copying over dependencies.

## Todos
- [x] Tests
    - Ensure all existing tests are passing
- [x] Documentation
    - Add something to README.md about testing Calico node
- [ ] ~~Release note~~

## Release Note
```release-note
None required
```

### CODE REVIEW

File: filesystem/sbin/start_runit [hunk 1]
```diff
@@ -13,4 +13,4 @@ fi
 # Export the nodename set by the startup procedure. 
 export NODENAME=$(cat /var/lib/calico/nodename)
 
-exec /sbin/runsvdir -P /etc/service/enabled
+exec /usr/bin/runsvdir -P /etc/service/enabled
```
Comment (stevegaossou) [inline filesystem/sbin/start_runit]: This changed due to switching to Debian, the default path for `runsvdir` is different on a Debian install.

File: filesystem/etc/rc.local [hunk 2]
```diff
@@ -58,12 +59,12 @@ esac
...
 # If running libnetwork plugin in a separate container, CALICO_LIBNETWORK_ENABLED would be false.
 # CALICO_LIBNETWORK_ENABLED is "false" by default. It can be set by passing `--libnetwork` flag while starting the calico/node via calicoctl
-if [ "$CALICO_LIBNETWORK_ENABLED" == "true" ]; then
+if [ "$CALICO_LIBNETWORK_ENABLED" = "true" ]; then
 	echo "Starting libnetwork service"
 	cp -a /etc/service/available/libnetwork  /etc/service/enabled/
 fi
 
-if [ "$CALICO_DISABLE_FILE_LOGGING" == "true" ]; then
+if [ "$CALICO_DISABLE_FILE_LOGGING" = "true" ]; then
```
Comment (stevegaossou) [inline filesystem/etc/rc.local]: These changes is a result of me debugging borked tests as a result of switching to Debian. 

I don't think these are strictly necessary (these ultimately were not the causes of the borked tests). However, running `shellcheck` on this (similar to what I did on [CNI Plugin base image refactor](https://github.com/projectcalico/cni-plugin/pull/708/files#diff-282edec51ec64b45bd3e29ddced5e78f)) showed the above recommended changes.

**Comment by stevegaossou** [comment]
All System Tests (`make ci`) are green now. ☝️ 

Only remaining blocker is to fix the k8s tests (`make k8s-test`) which is a much smaller batch.

File: Makefile [hunk 3]
```diff
@@ -538,7 +548,8 @@ st: dist/calicoctl busybox.tar calico-node.tar workload.tar run-etcd calico_test
 ###############################################################################
 .PHONY: ci
 ## Run what CI runs
-ci: static-checks fv image-all st
+# ci: static-checks fv image-all st
+ci:
```
Comment (stevegaossou) [inline Makefile]: This is debugging right now. Will be reverted once everything with `make k8s-test` is resolved.

**Comment by stevegaossou** [comment]
Omg, it's green!

**Comment by stevegaossou** [comment]
Finally ... 🙏

File: Makefile [hunk 4]
```diff
@@ -502,6 +506,12 @@ k8s-run-test: calico_test.created
 	    sh -c 'cp /root/.kubeadm-dind-cluster/kubectl /bin/kubectl && ls -ltr /bin/kubectl && which kubectl && cd /code/tests/k8st && \
 	           nosetests $(K8ST_TO_RUN) -v --with-xunit --xunit-file="/code/report/k8s-tests.xml" --with-timer'
 
+# Needed for Semaphore CI (where disk space is a real issue during k8s-test)
+.PHONY: remove-go-build-image
+remove-go-build-image:
+	@echo "Removing $(CALICO_BUILD) image to save space needed for testing ..."
+	@-docker rmi $(CALICO_BUILD)
+
```
Comment (stevegaossou) [inline Makefile]: Added this flag and target to specifically address issues with `k8s-test` failing on Semaphore CI due to 10 GB disk space hard limit. 

We were already riding that 10GB limit line before this base image change. After it, the size of calico node went 2x, and it pushed us over the edge. 

I configured the Semaphore CI job for `k8s-test` specific to use this flag to remove the Calico Go Build image right before the tests start, which saves us about 1.3GB of space. 

Screen Shot 2019-03-30 at 9 19 33 PM

Any running the targets locally, will not have their Go build image removed (as long as they don't pass that flag).

**Comment by stevegaossou** [comment]
This one is finally ready 4 review. 

cc @caseydavenport

File: README.md [hunk 5]
```diff
@@ -24,3 +24,106 @@ The code in this repository can be built and tested using the Makefile.
 For more information, see `make help`.
 
 [contrib]: https://github.com/projectcalico/calico/blob/master/CONTRIBUTING_CODE.md
+
+## How can I run Calico Node tests?
+
+Calico Node uses Python and [Nose](https://nose.readthedocs.io/en/latest/) for running tests.
```
Comment (caseydavenport) [inline README.md]: nit on the readme: "Calico Node" should be `calico/node` when referring to the image, and simply "node" (lowercase) when referring to the repository.

File: README.md [hunk 6]
```diff
@@ -24,3 +24,106 @@ The code in this repository can be built and tested using the Makefile.
...
+
+Calico Node tests are divided into the following categories: 
+- `fv`: ???
```
Comment (caseydavenport) [inline README.md]: fv: Runs package scoped ginkgo tests

File: README.md (hunk 5 above)
Comment (caseydavenport) [reply README.md]: Also note that we use golang / ginkgo for some of the tests (fv) - it's just the full integrations tests in python (which is most of them)

File: README.md [hunk 7]
```diff
@@ -24,3 +24,106 @@ The code in this repository can be built and tested using the Makefile.
...
+- This only applies to `st` tests 
+- A subset of the `st` are wrapped by `debug_failures(fn)` function found in `./tests/st/utils/utils.py`
+- You should be able to wrap whatever test you want 
+- Uses Python's `pdb.set_trace()` library function, allows you to halt executino and step into the containers involved in the test for debugging 
+3. Use manual breakpoints 
+- A more primitive approach is just to add your own breakpoints (using something like `time.sleep(x)`)
+- You should know where to add these after reviewing the diagnostic logs for failed tests (by looking at the stacktraces)
+
+## Linux Dependencies 
+Below is a listing of dependencies that Calico node has on the underlying kernel. The list is currently _partially_ complete.
```
Comment (caseydavenport) [inline README.md]: I'd tweak the wording here subtly - these are userspace tools packaged into the node container. Each of these userspace tools does however imply certain underlying kernel support.

**Comment by stevegaossou** [comment]
> This looks good to me - just a few minor README comments, and then if you could squash your commits I'll press the ol' merge button.

I did a rebase and force-pushed my stuff. I collapsed it into two separate commits (because my commits were interleaved with some other ones). Always feels odd to see commits that weren't from my feature branch, but it is what it is. I did dry-run test merge into `master` locally and it went okay. 

I'm still waiting on official 👍 before we merge this. I'll ping @caseydavenport once I get confirmation.

**Comment by caseydavenport** [comment]
@stevegaossou hrm, not sure how that happened - it shouldn't have any of the marvin commits in it. 

What process did you use to rebase?

**Comment by stevegaossou** [comment]
> @stevegaossou hrm, not sure how that happened - it shouldn't have any of the marvin commits in it.
> What process did you use to rebase?

I did a `git rebase -i HEAD~16` (where I had 16 commit originally on this branch). I think the problem was that I had done regular merge downs from `master` so when I when to collapse everything the history was interleaved with commits from `master`. 

Here's what it looked like: 
```
pick 30d2d89 Switching node base image from alpine to debian.
pick 6e36c2c Bumping up the version of Debian to Buster-slim (to avoid CVEs and ensure versions of ipset, iptables, etc match or are newer than alpine:3.8).
pick 4760af0 Trying to switch iptables 1.8.2 to use legacy mode instead of nftables mode (which is the new default).
pick 63fad39 Moving back to debian 9.8-slim, as buster-slim (even using legacy iptables commands) breaks the BGP tests (iptables-save errors out).
pick 7f52205 Switching back to debian:buster-slim, also adding kmod package to resolve problem with ip6tables-save.
pick aa560ba Adding another felix dependency.
pick 44fd179 Adding another felix dependency (this time to get arp).
pick d86cde4 Disabling CI tests temporarily to debug CI issues with k8s-tests.
pick 474c9e9 Automated pin updates
pick 6c4e343 Avoid apparent pyyaml vulnerability
pick 71979dd Automated pin updates
pick d9c840e Automated pin updates
pick 9a3ad99 Automated pin updates
pick 9b5046b Adding flag and target to remove Go build image from machine, needed for Semaphore CI (where disk space is a real issue during k8s-test).
pick 864f9ee Adding CI tests back.
pick 3a9e60b Adding some documentation for running tests for Calico Node.
pick b694f2d Adding in dependencies section to documentation.
pick 9ecda90 Making some corrections to documentation based on feedback.
pick 1ef3c1a Automated pin updates
pick e5fc2cb Stop printing log to stdout in main

# Rebase f8ee453..eb3751d onto f8ee453 (20 commands)
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
# d, drop = remove commit
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out
```

I have a backup version of this branch that is unchanged.

**Pull request merged**